apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,163 @@
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.classLoaderCanLoadClass;
+import static co.elastic.apm.agent.bci.bytebuddy.CustomElementMatchers.implementationVersionLte;
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+public class JavalinInstrumentation extends TracerAwareInstrumentation {
+
+    private static final String FRAMEWORK_NAME = "Javalin";
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("io.javalin.http.Handler")).and(not(isInterface()));
+    }
+
+    @Override
+    public ElementMatcher.Junction<ClassLoader> getClassLoaderMatcher() {
+        return classLoaderCanLoadClass("io.javalin.http.Handler");
+    }
+
+    // ctx.handlerType() was introduced in Javalin 3.13.8 and above, so instrument only in that case
+    @Override
+    public ElementMatcher.Junction<ProtectionDomain> getProtectionDomainPostFilter() {
+        return implementationVersionLte("3.13.7");
+    }
+
+    @Override
+    public ElementMatcher<? super MethodDescription> getMethodMatcher() {
+        return named("handle").and(takesArgument(0, named("io.javalin.http.Context")));
+    }
+
+    @Override
+    public Collection<String> getInstrumentationGroupNames() {
+        return Collections.singleton("javalin");
+    }
+
+    @Override
+    public String getAdviceClassName() {
+        return "co.elastic.apm.agent.javalin.JavalinInstrumentation$HandlerAdapterAdvice";
+    }
+
+    public static class HandlerAdapterAdvice {
+
+        @Advice.OnMethodEnter(suppress = Throwable.class, inline = false)
+        public static Object setSpanAndTransactionName(@Advice.This Handler handler,
+                                                       @Advice.Argument(0) Object ctxObject) throws Exception {
+            final Transaction transaction = tracer.currentTransaction();
+            if (transaction == null) {
+                return null;
+            }
+            Context ctx = ((Context) ctxObject);
+
+            final String handlerClassName = handler.getClass().getName();
+            final boolean isLambdaHandler = handlerClassName.equals("co.elastic.apm.agent.javalin.JavalinHandlerLambdaInstrumentation$WrappingHandler");
+
+            // transaction name gets only set if we are dealing with a HTTP method processing, not before/after handlers
+            if (ctx.handlerType().isHttpMethod()) {
+                final StringBuilder name = transaction.getAndOverrideName(PRIO_HIGH_LEVEL_FRAMEWORK);
```

Comment:
Shouldn't you use:
```suggestion
                final StringBuilder name = transaction.getAndOverrideName(PRIO_HIGH_LEVEL_FRAMEWORK, false);
```
So that you don't override the name for every handler?